src/main/java/io/cdap/plugin/gcp/spanner/sink/SpannerOutputFormat.java

```diff
@@ -231,4 +243,125 @@ public void abortTask(TaskAttemptContext taskAttemptContext) {
+          builder.set(name)
+            .toDateArray(
+              dateList.stream()
+                .map(date -> Date.fromYearMonthDay(date.getYear(), date.getMonthValue(), date.getDayOfMonth()))
+                .collect(Collectors.toList())
+            );
+          break;
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          Collection<ZonedDateTime> timestampList = toCollection(name, logicalType.toString(), values);
+          builder.set(name)
+            .toTimestampArray(
+              timestampList.stream()
+                .map(ts -> Timestamp.ofTimeSecondsAndNanos(ts.toEpochSecond(), ts.getNano()))
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + logicalType + "' logical type currently not supported.");
+      }
+    } else {
+      Schema.Type type = unionSchema.getType();
+      switch (type) {
+        case BOOLEAN:
+          Collection<Boolean> listBoolean = toCollection(name, type.toString(), values);
+          builder.set(name).toBoolArray(listBoolean);
+          break;
+        case STRING:
+          Collection<String> listString = toCollection(name, type.toString(), values);
+          builder.set(name).toStringArray(listString);
+          break;
+        case LONG:
+          Collection<Long> listValue = toCollection(name, type.toString(), values);
+          builder.set(name).toInt64Array(listValue);
+          break;
+        case DOUBLE:
+          Collection<Double> listDouble = toCollection(name, type.toString(), values);
+          builder.set(name).toFloat64Array(listDouble);
+          break;
+        case INT:
+          Collection<Integer> integerList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toInt64Array(
+              integerList.stream()
+                .map(Integer::longValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case FLOAT:
+          Collection<Float> floatList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toFloat64Array(
+              floatList.stream()
+                .map(Float::doubleValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case BYTES:
+          Collection<byte[]> bytesList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toBytesArray(
+              bytesList.stream()
+                .map(ByteArray::copyFrom)
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + type.name() + "' type currently not supported.");
+      }
+    }
+  }
+
+  private static <T> Collection<T> toCollection(String fieldName, String fieldType, Object value) {
```

Comment:
I assume this means Spanner doesn't allow arrays of arrays (otherwise this needs to recursively convert the collection contents).

Reply:
Yes, spanner doesn’t support Array of Arrays https://cloud.google.com/spanner/docs/data-types#array-type